[READY] Moved change tracking to python. Per buffer diagnostics

davits opened the pull request:
# PR Prelude

Thank you for working on YCM! :)


- [x] I have read and understood YCM's [CONTRIBUTING][cont] document.
- [x] I have read and understood YCM's [CODE_OF_CONDUCT][code] document.
- [x] I have included tests for the changes in my PR. If not, I have included a
  rationale for why I haven't.
- [x] **I understand my PR may be closed if it becomes obvious I didn't
  actually perform all of these steps.**

Adds an emulation of Vim buffer object and moves change tracking into python.
Now diagnostics are stored per buffer, which effectively solves #2165 .

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2631)
<!-- Reviewable:end -->

homu commented:
:umbrella: The latest upstream changes (presumably #2629) made this pull request unmergeable. Please resolve the merge conflicts.

bstaletic commented:
I've made a quick review of these changes. I like the buffer abstraction which at the first glance seems to make things easier to write/add/change. Also, tests are passing and brief testing tells me this works.

However, I don't know python well enough to make the clear judgement if this is LGTM, so for now, I'll leave that part to the others. Only two minor comments.

---

Reviewed 6 of 8 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-KkHI8y_KjbYRJaVH9Zq:b9aa4sa) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
> ```Python
>   def __missing__( self, key ):
>     value = self[ key ] = Buffer( key, self._user_options )
>     return value
> ```

Temporary object `value` is not needed. Could be written as
```py
return seelf[ key ] = Buffer( key, self._user_options )
```

---

*[python/ycm/diagnostic_interface.py, line 68 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkLvEfqyqnFXlko0sJo:-KkLvEfqyqnFXlko0sJp:b-8cnplj) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/diagnostic_interface.py#L68)):*
> ```Python
>     if not self._user_options[ 'always_populate_location_list' ]:
>       self._UpdateLocationList()
>     return bool( self._diagnostics )
> ```

I won't try to be smart here. Can you explain what's going on in thte new `PopulateLocationList()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-KkLzJFl_ovQm5l7915O:bcxq29a)*
<!-- Sent from Reviewable.io -->

davits commented:
Update:
Reverted changes from initial commits where the whole file parsing chain was moved into python, it was done because of new awesome way to handle parse requests with timers was merged.
I would recommend to set initial revision to the base while reviewing, it would be much easier in that way.

---

Review status: 1 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-Kkfe9hsJdW28p0Vegh0:bngk6z2) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

Temporary object `value` is not needed. Could be written as
```py
return seelf[ key ] = Buffer( key, self._user_options )
```
</blockquote></details>

Strangely for C++ oriented world that does not work :), flake throws an error.

---

*[python/ycm/diagnostic_interface.py, line 68 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkLvEfqyqnFXlko0sJo:-KkfeTCf9ssg9Isp75By:bb4vccg) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/diagnostic_interface.py#L68)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

I won't try to be smart here. Can you explain what's going on in thte new `PopulateLocationList()`?
</blockquote></details>

Basically it's the same thing. I just moved 'always_populate...' check here from youcompleteme.py , and did a little refactoring.
There is a bigger change in the queue for the diagnostic_interface, this is just a small fraction of it, which I couldn't cut out from this change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-Kkfffi9Agwf52bzvfsU:b-ua58d8)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Reviewed 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-KkhYWcwA9D31LBvLkBt:b-d5o7sf) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Strangely for C++ oriented world that does not work :), flake throws an error.
</blockquote></details>

Hmm, that is strange. I see no reason why it wouldn't work. Unfortunately, I'm not at my pc right now, so I can't check. 

Could you add a comment explaining why the temporary object is needed? I can see my future self trying to remove it just to be greeted by an error.

Also would this work?
```py
self[ key ] = Buffer( key, self._iser_options )
return self[ key ]
```

---

*[python/ycm/youcompleteme.py, line 583 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkhWm6AAwa4hjkfwgyy:-KkhWm6BoC2c0wsoRUCV:bigtvel) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L583)):*
> ```Python
>       # Note: it is the server's responsibility to determine the frequency of
>       # error/warning/prompts when receiving a FileReadyToParse event, but
>       # it our responsibility to ensure that we only apply the
> ```

Typo here. Should either be `# it is` or `# it's`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
With my two comments resolved this is <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-Kl2wk0Wu0x3zwkmUObF:b-85t60x)*
<!-- Sent from Reviewable.io -->

vheon commented:
I did a quick read of the code; I also like the Buffer abstraction but I left a couple of questions before continuing with the review :confused: 

---

Reviewed 1 of 7 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-Kl5W3VF2vwnFotjYcf_:b-hzo642) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Hmm, that is strange. I see no reason why it wouldn't work. Unfortunately, I'm not at my pc right now, so I can't check. 

Could you add a comment explaining why the temporary object is needed? I can see my future self trying to remove it just to be greeted by an error.

Also would this work?
```py
self[ key ] = Buffer( key, self._iser_options )
return self[ key ]
```
</blockquote></details>

IMO the latest suggestion by @bstaletic reads better but I don't have a strong opinion on this :+1: 

---

*[python/ycm/vimsupport.py, line 595 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5Z9h3_PyKe8VTBMQK:-Kl5Z9h3_PyKe8VTBMQL:b-kbu23z) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/vimsupport.py#L595)):*
> ```Python
> 
> 
> def GetBufferFiletypes( bufnr ):
> ```

I don't see any usage of this function.

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-Kl5_UD17OMS5AixNENn:b-tj3vbe) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
> ```Python
> 
>   def _GetCurrentBuffer( self ):
>     return self._buffers[ vimsupport.GetCurrentBufferNumber() ]
> ```

we call `_GetCurrentBuffer` basically every time we have to do something which means that we look up in the buffers table every time we have to do something. Wouldn't be better to cache the current buffer and just set it on `BufEnter`? Just an idea :)

---

*[python/ycm/youcompleteme.py, line 371 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_6-OjSkofUFj2V6m:-Kl5_6-OjSkofUFj2V6n:ba88wdu) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L371)):*
> ```Python
>       return
> 
>     if not self.IsServerReadyWithCache():
> ```

I'm not understanding why is this needed now while it was not before :confused: Are we addressing other issue other than the per-buffer diagnostic and changetick moving to Python?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-Kl5aur-yXYyZ27sD_N-:bst9eym)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-Kl8ZzkKaD7i4p_RhIJM:b8yqaqu) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

we call `_GetCurrentBuffer` basically every time we have to do something which means that we look up in the buffers table every time we have to do something. Wouldn't be better to cache the current buffer and just set it on `BufEnter`? Just an idea :)
</blockquote></details>

While caching an often used value is often a good idea, we need to be careful here. What happens if a user manually changes the filetype? I've done that before. Vim assumes that `.h` file extension means C++. So I've manually changed it to C to get the right snippets from Ultisnips through YCM. I agree that this is a rare case, but it is something we need to consider. I believe that it could be a common thing in codebases that mix C and C++.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-KlAcLx1TqOYGQKJDGYo:bkx9mhl) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

While caching an often used value is often a good idea, we need to be careful here. What happens if a user manually changes the filetype? I've done that before. Vim assumes that `.h` file extension means C++. So I've manually changed it to C to get the right snippets from Ultisnips through YCM. I agree that this is a rare case, but it is something we need to consider. I believe that it could be a common thing in codebases that mix C and C++.
</blockquote></details>

@bstaletic I'm not sure I understood you. What I meant is that we're continuously getting the number of the current buffer and getting the Buffer object from the map. Even if you change the filetype of a buffer its number doesn't change :confused:  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-KlBRLYs1LOfwiLjQr1Q:b-flwkxi) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

@bstaletic I'm not sure I understood you. What I meant is that we're continuously getting the number of the current buffer and getting the Buffer object from the map. Even if you change the filetype of a buffer its number doesn't change :confused:  
</blockquote></details>

@vheon Forget my comment. It was utter nonsense as I was thinking of keeping track of filet type instead of buffer number.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 362 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KlELxCkMcyS9X38qONo:-KlELxCkMcyS9X38qONp:b-ml1hvc) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L362)):*
> ```Python
> 
> 
>   def NeedsReparse( self ):
> ```

I've noticed this function is not covered and then I wasn't able to find where it is used. So are we even using this?

This, along with `GetBufferFiletypes` that @vheon mentioned, make up 4 out of 5 lines that are not covered by tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 362 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KlELxCkMcyS9X38qONo:-KlFBi3dmYkvQtQnWiTo:bvkpwj4) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L362)):*
<details><summary><i>Previously, bstaletic wrote…</i></summary><blockquote>

I've noticed this function is not covered and then I wasn't able to find where it is used. So are we even using this?

This, along with `GetBufferFiletypes` that @vheon mentioned, make up 4 out of 5 lines that are not covered by tests.
</blockquote></details>

This is called from the VimL layer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[python/ycm/youcompleteme.py, line 362 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KlELxCkMcyS9X38qONo:-KlFDkjdxYh6ZWqyUKhP:b-fo43mo) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L362)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

This is called from the VimL layer.
</blockquote></details>

Righ. My bad. I was only checking the python layer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

davits commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-KlKS8IPCIKhDCdBaWDT:b6a58vy) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

IMO the latest suggestion by @bstaletic reads better but I don't have a strong opinion on this :+1: 
</blockquote></details>

The @bstaletic 's suggestion does double lookup for the key.
Maybe mine too, I am not sure now. python is a strange animal :)

---

*[python/ycm/vimsupport.py, line 595 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5Z9h3_PyKe8VTBMQK:-KlKKNgxCMGy0a_qLNHi:ba3dfob) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/vimsupport.py#L595)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

I don't see any usage of this function.
</blockquote></details>

Removed

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-KlKEFMdp17AyAKgZB7V:bqcm0oi) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

@vheon Forget my comment. It was utter nonsense as I was thinking of keeping track of filet type instead of buffer number.
</blockquote></details>

The problem with `BufEnter` is that it is not called at startup while server is becoming ready. OnCursorMoved, Error/WarningCount ... function will fail at startup.
In the initial iterations the current buffer was being cached in the OnFileReadyToParse function, but that had problems of its own.

But I can make it work with `BufEnter` by adding more `IsServerReadyWithCache` checks.

---

*[python/ycm/youcompleteme.py, line 371 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_6-OjSkofUFj2V6m:-KlKPnQPYQtRdhtm1Txr:bp5br72) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L371)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

I'm not understanding why is this needed now while it was not before :confused: Are we addressing other issue other than the per-buffer diagnostic and changetick moving to Python?
</blockquote></details>

There is no issue, I am trying to block parse requests while server is not ready yet. To prevent server spamming and to not let buffer be marked as parsed, i.e. NeedsReparse returning False.

---

*[python/ycm/youcompleteme.py, line 583 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkhWm6AAwa4hjkfwgyy:-KlKDXj8NrwfXKzLR7Is:b-896fix) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L583)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Typo here. Should either be `# it is` or `# it's`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-KlLKMskJMb0yiGbGFxj:b89xfxn) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

The @bstaletic 's suggestion does double lookup for the key.
Maybe mine too, I am not sure now. python is a strange animal :)
</blockquote></details>

I tried to benchmark these two approaches. I couldn't draw a definitive conclusion, even though @davits's suggestion seems logical to be faster, as it indeed uses only one lookup. It is possible that I couldn't do it because I was using a much simpler dictionary than the one used here in an attempt to simplify my test.

Considering python's overall performance and the fact that some people (me included) sometimes open a lot of files I'm now in favour of @davits's solution.

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-KlLM1tEx4NyPv0Yq7Hj:b-my04ya) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

The problem with `BufEnter` is that it is not called at startup while server is becoming ready. OnCursorMoved, Error/WarningCount ... function will fail at startup.
In the initial iterations the current buffer was being cached in the OnFileReadyToParse function, but that had problems of its own.

But I can make it work with `BufEnter` by adding more `IsServerReadyWithCache` checks.

</blockquote></details>

Can we be sure the extra `IsServerReadyWithCache`checks is a penalty worth taking in exchange for having cached values? Maybe we should see some numbers before changing if this works properly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I think this is <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from me. @davits Good work!

Since @vheon has started but hasn't finished his review, not triggering zzbot. Also, no `[READY]` tag.

@davits You might want to change the PR title tag to `[READY]` once you're ready to have this merged.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-KlMelxydqRC_97ohmme:bf54hz7)*
<!-- Sent from Reviewable.io -->

davits commented:
Thanks.

As soon as @vheon and I are agreed on `IsServerReadyWithCache()` vs `_GetCurrentBuffer()` dilemma.

@vheon I can cache current buffer on `OnBufferEnter` and add more checks for server being ready or keep `_GetCurrentBuffer()` and remove `IsServerReadyWithCache()` calls.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[python/ycm/buffer.py, line 100 at r2](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-KkHI8y_KjbYRJaVH9Zp:-KlNX3JKyIMr5jGXtLaH:b-6qzdcn) ([raw file](https://github.com/valloric/youcompleteme/blob/9024946cfd7fd29e9a57a99c394018d6161b62de/python/ycm/buffer.py#L100)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I tried to benchmark these two approaches. I couldn't draw a definitive conclusion, even though @davits's suggestion seems logical to be faster, as it indeed uses only one lookup. It is possible that I couldn't do it because I was using a much simpler dictionary than the one used here in an attempt to simplify my test.

Considering python's overall performance and the fact that some people (me included) sometimes open a lot of files I'm now in favour of @davits's solution.
</blockquote></details>

Thanks @bstaletic

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-KlNU7i0_9XAplrfpZIF:b-iiggxx) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Can we be sure the extra `IsServerReadyWithCache`checks is a penalty worth taking in exchange for having cached values? Maybe we should see some numbers before changing if this works properly.
</blockquote></details>

`IsServerReadyWithCache()` has lower cost then `_GetCurrentBuffer()`, boolean check vs binary search.
I think we are all trying to be perfectionists here, which I like :)
IMHO in both cases performance impact is negligible, so it sums up to the question which style do we prefer ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-KlNX744ksC5G4scsHWE:b-6ptd0u)*
<!-- Sent from Reviewable.io -->

davits commented:
Review status: 3 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[python/ycm/youcompleteme.py, line 136 at r4](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-Kl5_UD17OMS5AixNENm:-KlmA3-g42PVhNt9rL3E:b-tm3by2) ([raw file](https://github.com/valloric/youcompleteme/blob/b6bbe6388fa29f3be2b248ac5ecfb3c022e7bcdf/python/ycm/youcompleteme.py#L136)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

`IsServerReadyWithCache()` has lower cost then `_GetCurrentBuffer()`, boolean check vs binary search.
I think we are all trying to be perfectionists here, which I like :)
IMHO in both cases performance impact is negligible, so it sums up to the question which style do we prefer ?
</blockquote></details>

Added caching of current buffer on `BufEnter`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I'm not firing `zzbot` because there is no `READY` tag yet :)

---

Reviewed 2 of 3 files at r5, 2 of 5 files at r6.
Review status: 5 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-KlnpesZOhtgHNKTKa_8:b-z36xii)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
@zzbot r+

---

Reviewed 5 of 5 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2631#-:-Klrl4kYJUHpg-calClB:bi59kwd)*
<!-- Sent from Reviewable.io -->
